experimental/Geometry/Schemes.jl

```diff
@@ -0,0 +1,1244 @@
+
+Return the affine scheme corresponding to the ring $R/(0)$.
+"""
+function Spec( R::T ) where{T <: MPolyRing}
+  I = ideal(R, zero(R))
+  k = coefficient_ring( R )
+  return Spec(k, R, I )
+end
+
+function Spec( R::T, I::MPolyIdeal{U} ) where{ T <: MPolyRing, U <: MPolyElem }
+  k = coefficient_ring(R)
+  return Spec(k, R, I )
+end
+
+# Construct affine n-space over the ring k.
+function affine_space( k::Ring, n::Int, name::String="x" )
+  R, x = PolynomialRing( k, name => (1:n))
+  return Spec( R )
+end
+
+function Base.show( io::Base.IO, X::AffineScheme )
+  if isdefined( X, :name )
+    Base.print( io, X.name )
+    return
+  end
+  Base.print( io, "Affine scheme over " )
+  Base.print( io, base_ring(X) )
+  Base.print( io, "\n" ) 
+  Base.print( io, "given as the quotient of the polynomial ring \n" )
+  Base.print( io, ambient_ring(X) )
+  Base.print( io, "\nby the ideal \n" )
+  Base.print( io, defining_ideal(X) )
+end
+
+
+function Base.show( io::Base.IO, X::SpecPrincipalOpen)
+  Base.print( io, "Principal open subscheme of \n" )
+  Base.print( io, root(X) )
+  Base.print( io, "\n" )
+  Base.print( io, "localized at\n" )
+  Base.print( io,  denoms(X))
+end
+
+############################################################################
+# Morphisms of affine schemes.
+#
+# These need to take into account both the Specs and their localizations. 
+# The information on the morphism is stored either as an explicit 
+# homomorphism of polynomial rings for the coordinate rings of the 
+# associated specs, or as a list of rational functions, the images of 
+# the variables. 
+
+mutable struct AffSchMorphism{S,Tdom, Udom, Tcod, Ucod}
+  domain::AffineScheme{S, Tdom, Udom}
+  codomain::AffineScheme{S, Tcod, Ucod}
+
+  # Variables for caching 
+  #
+  # The morphism must be given either in terms of an algebra homomorphism 
+  # or in terms of fractions. Here we exploit the fact that to give a morphism 
+  # of localized schemes, it is only necessary to prescribe the images of 
+  # the generators of the ambient polynomial ring of the root.
+  pullback::AlgHom
+  imgs_frac::Vector{Frac{Ucod}}
+  inverse::AffSchMorphism{S,Tdom,Udom,Tcod,Ucod} # optional field in case this is indeed an isomorphism
+  sections::Vector{AffSchMorphism{S,Tdom,Udom,Tcod,Ucod}} # optional field for storing sections
+
+  function AffSchMorphism( domain::AffineScheme{S,Td,Ud},
+                           codomain::AffineScheme{S,Tc,Uc}, pullback::AlgHom
+                           ) where {S,Td,Ud,Tc,Uc}
+
+   if base_ring(domain) != base_ring(codomain)
+      error( "the base rings of the domain and the codomain do not coincide!" )
+    end
+    k = base_ring(domain)
+    #if domain(pullback) != ambient_ring(R) || codomain(pullback) != domain.R
+      #error( "the domain and codomain of the given ring homomorphism is not compatible with the affine schemes." )
+    #end
+    x = new{S,Td,Ud,Tc,Uc}()
+    x.domain = domain
+    x.codomain = codomain
+    x.pullback = pullback
```

Comment:
```suggestion
    x = new{S,Td,Ud,Tc,Uc}(domain, codomain, pullback)
```

Reply:
That's also how I did this originally. The format used here I learned from Simon. Is there a performance reason to favor the one-liner?

Reply:
It's more idiomatic, and more concise.